presto-hive/src/main/java/io/prestosql/plugin/hive/util/HiveBucketing.java

```diff
@@ -98,6 +102,64 @@ public static int getHiveBucket(BucketingVersion bucketingVersion, int bucketCou
         return getBucketNumber(getBucketHashCode(bucketingVersion, types, values), bucketCount);
     }
 
+    private static Optional<Set<Integer>> getHiveBuckets(BucketingVersion bucketingVersion, int bucketCount, List<TypeInfo> types, List<List<NullableValue>> values)
+    {
+        long explorationCount;
+        try {
+            explorationCount = values.stream()
+                    .mapToLong(List::size)
+                    .reduce(1, Math::multiplyExact);
+        }
+        catch (ArithmeticException e) {
+            return Optional.empty();
+        }
+        long explorationLimit = Math.max(bucketCount * BUCKETS_EXPLORATION_LIMIT_FACTOR, BUCKETS_EXPLORATION_GRANTED_LIMIT);
+        if (explorationCount > explorationLimit) {
+            return Optional.empty();
+        }
+
+        Set<Integer> buckets = new HashSet<>();
+        getHiveBucketsByDiscreteSets(bucketingVersion, bucketCount, types, values, buckets, 0, 0);
+        if (buckets.size() < bucketCount) {
+            return Optional.of(ImmutableSet.copyOf(buckets));
+        }
+        return Optional.empty();
+    }
+
+    private static void getHiveBucketsByDiscreteSets(
+            BucketingVersion bucketingVersion,
+            int bucketCount,
+            List<TypeInfo> types,
+            List<List<NullableValue>> values,
+            Set<Integer> buckets,
+            int hashedPrefix,
+            int index)
+    {
+        if (index < values.size()) {
+            TypeInfo type = types.get(index);
+            for (NullableValue value : values.get(index)) {
+                int newHashedPrefix;
+                switch (bucketingVersion) {
```

Comment:
Make these instance methods on the emum

Reply:
I applied the same pattern as in `getBucketHashCode()` methods, for style consistency.

Reply:
We should fix those as well.